### PR TITLE
Modify approximate view factor calculation logic

### LINKED ISSUE #8976: Incorrect logic in approximate view factor calculation
Issue overview
--------------
Current logic in [CalcApproximateViewFactors](https://github.com/NREL/EnergyPlus/blob/1a8e4918d633d44c846d7b965c11afb849692111/src/EnergyPlus/HeatBalanceIntRadExchange.cc#L1345) function for determining what surfaces can "see" each other is shown below.  The documentation lists the conditions for surfaces to "see" themselves and the code doesn't appear to account for all of them.  Also, if Azimuth(i) = 359 and Azimuth(j) = 1, the two surfaces should be within the "SameAngleLimit", but they won't be.

Source Code:
```
                if ((state.dataSurface->Surface(SPtr(j)).Class == SurfaceClass::IntMass) ||
                    (state.dataSurface->Surface(SPtr(j)).Class == SurfaceClass::Floor) ||
                    (state.dataSurface->Surface(SPtr(j)).Class == SurfaceClass::Roof &&
                     state.dataSurface->Surface(SPtr(i)).Class == SurfaceClass::Floor) ||
                    ((std::abs(Azimuth(i) - Azimuth(j)) > SameAngleLimit) ||
                     (std::abs(Tilt(i) - Tilt(j)) >
                      SameAngleLimit)))
```

Documentation:
```
To summarize, using the Surface “Class”, the approximate view factors have:

1. No surface sees itself.
2. No Floor sees another floor.
3. All other surface types see Internal Mass.
4. All other surface types see floors.
5. Floors always see ceilings.
6. Floors always see roofs.
7. All other surfaces whose tilt or facing angle differences are greater than 10 degrees see each other.

If geometry is correct, conditions 1, 3, and 7 should take care of all surfaces, but the other conditions supply common sense when the geometry is incorrect. More information about the EnergyPlus view factor calculation is contained in the Engineering Reference document.
```

### Details
Some additional details for this issue (if relevant):
 - EnergyPlus 9.5

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by matthew-larson**
Pull request overview
---------------------
 - Fixes #8976 
 - This pull request addresses disconnects between the documentation and the source code logic for when two surfaces "see" each other for view factor calculations.  The documentation states 7 conditions for surfaces:

```
To summarize, using the Surface “Class”, the approximate view factors have:

1. No surface sees itself.
2. No Floor sees another floor.
3. All other surface types see Internal Mass.
4. All other surface types see floors.
5. Floors always see ceilings.
6. Floors always see roofs.
7. All other surfaces whose tilt or facing angle differences are greater than 10 degrees see each other.
```

The pull request adds the missing conditions to the if statement for the view factor calculation, including two floor surfaces and a roof and floor surface.  It was also found that surfaces with azimuth's less than 10 degrees but on opposite sides of the reference vector where being included in the calculation.  A unit test was created to test each of these conditions specifically.

**The large number of diffs are due to the original logic ignoring the second rule of no floor sees another floor.  Any surface saw a floor surface, even if it was a floor surface itself.  So any models with zones that contained multiple floors (plenums for example) or models with multiple stories that had surfacing or zone matching, the floors were seeing too many surfaces and thus resulted in incorrect view factors based on the constraints specified in the documentation.**

We do recognize that some of these conditions seem redundant or can be simplified based on other checks regarding the geometry and surface class but is probably for a bigger discussion at another time.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally

### CODE REVIEW

**Comment by mjwitte** [comment]
@matthew-larson Is this ready for review? Please merge in develop.

**Comment by matthew-larson** [comment]
@mjwitte Ready for review and merged in develop.

**Comment by Myoldmopar** [comment]
Is there any analysis of the diffs available here?  This seems to be causing diffs in lots and lots of files and while I appreciate the fix, I am hesitant to merge in something with so much change without lots of justifying data, tables, plots, etc.

**Comment by matthew-larson** [comment]
I spot-checked some models but can put together a detailed analysis to present here.

**Comment by nealkruis** [comment]
The primary analysis actually occurred through the unit test that was carefully designed to cover all the criteria listed in the documentation. It was very clear that the logic in the code was not consistent. We also found that several of the conditions in the documentation are redundant due to reciprocity rules (e.g., criteria 5 "Floors always see ceilings", and criteria 6 "Floors always see roofs" are implied by criteria 4 "All other surface types see floors"). That allowed us to simplify the updated logic a bit.

The changes are fairly isolated to the view factor calculations, but they cause big diffs in 380 files. As @matthew-larson mentioned, we spot checked the models to confirm that they contained situations where the logic was flawed, but there were a lot of them.

@Myoldmopar or @mjwitte do you have any suggestions of what plots/tables we could look at to help build confidence? The EIO generates a view factor report and we could see how that changes for some models, but I don't know if it's practical to do it for all of them.

**Comment by mjwitte** [comment]
@nealkruis Let's start out with 5ZoneAirCooled since many other files are based on the same geometry. Adding the view factors report and showing what the diffs are (I expect only the plenum zone would have diffs in view factors) and perhaps a plot of plenum air temperature for winter and summer design days to show the impact and maybe a plot of one of the floor surface inside face temps (which I would expect to be closer to the roof temp now). And maybe confirm a few of the files that don't show diffs are because they only have single floor surfaces in each zone (and no plenum).

**Comment by mjwitte** [comment]
And a few examples of the percent change in annual energy use would be useful (one 5Zone..., an ASHRAE..., and a RefBldg..., for example).

**Comment by matthew-larson** [comment]
First looking at the 5ZoneAirCooled model, I looked at the view factors report and the zone temperatures for comparison.  The view factors report showed a difference in the Surface View Factor Check Values for the PLENUM-1 zone, and only this zone.  The comparison is shown below (Beyond Compare) with the Develop results on top and the PR results on bottom and the PLENUM-1 zone highlighted in red with the differences across all values.

image

Below is a table looking at the PLENUM-1 zone temperature for Develop and the PR.  The winter design day maintains a constant design temperature so there is no change in outdoor air or PLENUM-1 zone temperature.  The temperature change is quite small.

image

The below table shows the comparison of the roof and two different floor surface inside face temperatures.  Again, very minor impact but does increase the roof surface inside face temperature slightly in the winter while decreasing the floor surface inside face temperatures.  Generally, the opposite occurs during the summer design day.

image

All of the 5Zone* models have these diffs in them while the 1Zone* models and others I checked that do not have multiple floors in a single zone do not show any diffs.

The difference in energy use for the 5ZoneAirCooled model was 0.0121% for January electricity, 0.0808% for July electricity, 0.0377% for January natural gas, and 0% for July natural gas.

The difference in energy use for the ASHRAE9012016_Hospital_Denver model was 0.0180% for January electricity, 0.0048% for July electricity, 0.0061% for January natural gas, and 0.0051% for July natural gas.

The difference in energy use for the RefBldgLargeOfficeNew2004_Chicago model was 0.0083% for January electricity, 0.0012% for July electricity, 0.0112% for January natural gas, and 0% for July natural gas.

**Comment by rraustad** [comment]
This is interesting:

    FinalCheckValue = FixedCheckValue = CheckConvergeTolerance = std::abs(sum_FixedF - N);

So the numbers shown for the [eio diffs](https://github.com/NREL/EnergyPlus/pull/8979#issuecomment-917100506) are convergence number? 9E-4 reducing to 6E-4. And this fix supposedly corrects a view factor issue with floors seeing other floors (or I guess surfaces in the same plane). Are there no eio reports for view factor? And if so, then why did those not change? at least for floors.

**Comment by nealkruis** [comment]
@rraustad there is a View Factor report in the EIO, but it has to be explicitly requested in the IDF.

**Comment by rraustad** [comment]
Then please request it and show the diffs so we can get comfortable with this.

**Comment by nealkruis** [comment]
@matthew-larson showed this in his post.

**Comment by rraustad** [comment]
I don't see that post on this thread, can you give me a link?

**Comment by nealkruis** [comment]
Or maybe not. That's just the view factor check for the zone.

**Comment by rraustad** [comment]
If there are no example files with view factors reported would it be a good idea to modify one file so there is a record for future diffs? I would say yes.

**Comment by matthew-larson** [comment]
Sorry about that, realized I only got the View Factor Info, not the View Factor Details.  Below show the before/after of this output report for PLENUM-1 zone.  The View Factor Details for all the other zones are identical for before/after.  I also added the View Factor Details output to the 5ZoneAirCooled.idf model.

Before:
image

After:
image

**Comment by mjwitte** [comment]
@matthew-larson Interesting results. The floor-to-floor view factors changed as expected (now zeros). And the wall view factors make more sense now, but I wasn't expecting such a big change (e.g. WALL-1PF to TOP-1 was 0.8742 and it now 0.4914). The new results makes more sense than the old, but can you explain  what fixed that?

**Comment by Myoldmopar** [comment]
This one probably has me the most concerned.  It's one of those PRs where it looks like it is improving things, but we are just a couple days away from freeze and this is causing regressions in **many** files.  I don't want to minimize the point that it could be improving things.  But I am super hesitant to drop this in without some form of widespread analysis on the results.  With big ESO diffs in 314 files, I need to see statistics on how much energy changed in the files, as well as how it affected component sizing results, etc.  As good as this change possibly is, this will need a big effort over the next couple days to have a chance at getting in.  I appreciate the work @matthew-larson and the reviews @mjwitte @rraustad.

**Comment by matthew-larson** [comment]
Thank you all for assisting in getting this reviewed and checked thoroughly.  I took the 5ZoneAirCooled.idf model and ran through the `CalcApproximateViewFactors` function, which results in the Approximate or User Input ViewFactors, and the first iteration of the `FixViewFactors` function, which is how the Final ViewFactors are calculated.  Both of these results are shown below.  

Before:
image

After:
image

The Approximate or User Input ViewFactors are exactly as expected based on the changes with the Wall and Roof surfaces matching Before/After, but the Floor surfaces no longer seeing each other in After and the view factors changing accordingly with the new "seen" area.  Going through the `FixViewFactors` function, it uses the approximate view factors and enforces reciprocity by averaging AiFij and AjFji (equation: 0.5 * (AF + transpose(AF))).  This is where the reduced number of seen surfaces effects the Wall and Roof surfaces.  With the increased view factors between the walls the floors, it decreases the view factors between the walls themselves, but also a significant increase between the walls and roof since the roof accounts for almost half of the "seen" area.  Below I have a screenshot of the manual calculations I did going through the process in the  `FixViewFactors` function to see the effects of the new Approximate ViewFactors for one of the walls (WALL-1PF), one of the floors (C1-1P), and the roof (TOP-1).  The [First Iteration] FixedF(j, i) line is the result after the first iteration, which doesn't converge, followed by the Final View Factor Reported line which shows the final view factor shown in the output to see the effects of the iterations to convergence.  Below the spreadsheet I've included the iterative process in the `FixViewFactors` function.  I attempted to show the steps and calculations in the process in the spreadsheet but if anyone has any questions or wants any other info or details on any aspect of this, let me know.

image

```
            for (i = 1; i <= N; ++i) {
                // Determine row coefficients which will enforce closure.
                Real64 const sum_FixedAF_i(sum(FixedAF(_, i)));
                if (std::abs(sum_FixedAF_i) > 1.0e-10) {
                    RowCoefficient(i) = A(i) / sum_FixedAF_i;
                } else {
                    RowCoefficient(i) = 1.0;
                }
                FixedAF(_, i) *= RowCoefficient(i);
            }

            //  Enforce reciprocity by averaging AiFij and AjFji
            FixedAF = 0.5 * (FixedAF + transpose(FixedAF));

            //  Form FixedF matrix
            for (i = 1; i <= N; ++i) {
                for (j = 1; j <= N; ++j) {
                    FixedF(j, i) = FixedAF(j, i) / A(i);
                    if (std::abs(FixedF(j, i)) < 1.e-10) {
                        FixedF(j, i) = 0.0;
                        FixedAF(j, i) = 0.0;
                    }
                }
            }

```

**Comment by matthew-larson** [comment]
I put together a spreadsheet (attached) where I pulled the relative differences for the system capacities, system flow rates, and meter results (System Capacities tab and Energy Enduses tab respectively) for the majority of the RefBldg* models.  The spreadsheet also still has the surface temperature data and view factor calculations if you care to see those as well.  

The maximum system capacity/flow rate difference in all the models is 0.72% in the mid-rise apartment and supermarket for only a system or two.  Regarding the energy enduses, the mid-rise apartment does have a 6.28% difference in the sum of the heating energy use, with all other models having less than 0.3% difference in sum.  Since the mid-rise apartment stood out the most, I looked into the view factors and the difference is that the internal mass surfaces were not always "seeing" all other surfaces.  Even-though the surfaces would "see" the internal mass surface, the internal mass surface would see the other surface and when the reciprocity was enforced by averaging AiFij and AjFji, the result would be different than if they both saw each other.  Since there is a high ratio of internal mass in this building, in every zone, relative to others, the effects were greater.
image

@Myoldmopar @mjwitte @rraustad Let me know if there are any additional questions or results you would like to see.  Thanks again for all your assistance with this one!

[View Factor PR Results.xlsx](https://github.com/NREL/EnergyPlus/files/7164733/View.Factor.PR.Results.xlsx)

**Comment by rraustad** [comment]
I have a mind to merge this just based on sheer effort. Or at least give an award for most detailed PR review presentation. Of course, it would also be nice to at least review these results. Maybe a short discussion after the iteration call?

File: src/EnergyPlus/HeatBalanceIntRadExchange.cc [hunk 1]
```diff
@@ -1463,22 +1461,24 @@ namespace HeatBalanceIntRadExchange {
...
+                //  No surface sees itself and no floor sees another floor
+                if (i == j || (state.dataSurface->Surface(SPtr(j)).Class == SurfaceClass::Floor &&
+                               state.dataSurface->Surface(SPtr(i)).Class == SurfaceClass::Floor))
+                    continue;
+                // All surface types see internal mass
+                // All surface types see floors
+                // Floors always see ceilings/roofs
+                // All other surfaces whose tilt or facing angle differences are greater than 10 degrees see each other
                 if ((state.dataSurface->Surface(SPtr(j)).Class == SurfaceClass::IntMass) ||
+                    (state.dataSurface->Surface(SPtr(i)).Class == SurfaceClass::IntMass) ||
```
Comment (nealkruis) [inline src/EnergyPlus/HeatBalanceIntRadExchange.cc:1476]: @mjwitte FYI, this is what was missing previously. All surfaces would see internal mass, but internal mass wouldn't see other surfaces.

**Comment by matthew-larson** [comment]
Below are the annual energy enduse results for the RefBldgMidriseApartmentNew2004_Chicago model, which had shown the most difference when looking at the hourly energy use results for the design day runs we reviewed on Wednesday.  Based on the annual energy use, the only reason the heating electricity relative difference was so high was because of it's minimal total energy use.  Electric unit heaters are used to serve the corridors.  The annual energy use shows  0.048% difference in natural gas use and 0.024% energy use difference in total energy use.  I've also included the Final View Factor report comparison showing how G N1 APARTMENT INTERNAL MASS has an initial approximate view factor of 0 for the G NWALL N1A wall and GWINDOW9 window surfaces.  This gets passed into the `FixViewFactors` function to determine the final view factors, which leads to incorrect final view factors for all surfaces.  So any model with internal mass may also exhibit the same incorrect initial approximate view factors.

**Design Day Run Hourly Results**
image

**Annual Enduse Energy Results**
image

**Final View Factor Comparison**
image

**Comment by matthew-larson** [comment]
Here is the same analysis done for the AirflowNetwork_MultiZone_House model:

**Design Day Run Hourly Results**
image

**Annual Enduse Energy Results**
image

**Approximate View Factor Comparison**
image

**Final View Factor Comparison**
image

**Comment by matthew-larson** [comment]
Here is the same analysis done for the 5ZoneAirCooled model:

**Design Day Run Hourly Results**
image

**Annual Enduse Energy Results**
image

**View Factor Comparison**
image

**Comment by Myoldmopar** [comment]
I have some uncertainty about this, but I have decided to merge it.  @matthew-larson thanks for all the work you've done to demonstrate the changes are an improvement and an obvious fix.  The energy impact seems relatively small, and I can't put into words what I would like to see further to actually feel better about it.  It's just a big-ish change.  These happen from time to time and to continue to improve the code, these have to happen.  This is nothing compared to some previous work (plant rewrites @EnergyArchmage ...) that had to deal with large diffs being justified and accepted.

Anywho, I'm merging this in so this fix can get into the users hands sooner rather than later.  Thanks all!

**Comment by matthew-larson** [comment]
Thanks @Myoldmopar!

**Pull request merged**